src/racer/util.rs

```diff
@@ -351,54 +362,142 @@ fn check_rust_sysroot() -> Option<path::PathBuf> {
-            // RUST_SRC_PATH is considered. This should either expect a single
-            // path to be provided, or all paths should be considered. The
-            // latter option would need to be supported in the rest of racer.
-
-            // Unwrap is ok here since split returns the original string
-            // even if it doesn't contain the split pattern.
-            let v = srcpaths.split(nameres::PATH_SEP).next().unwrap();
-            let f = path::Path::new(v);
-            if !f.exists() {
-                Err(RustSrcPathError::DoesNotExist(f.to_path_buf()))
-            } else if !f.join("libstd").exists() {
-                Err(RustSrcPathError::NotRustSourceTree(f.join("libstd")))
-            } else {
-                Ok(())
-            }
-        },
-        _ => {
-            if let Some(path) = check_rust_sysroot() {
-                env::set_var("RUST_SRC_PATH", path);
-                Ok(())
-            } else {
-                let default_paths = [
-                    "/usr/local/src/rust/src",
-                    "/usr/src/rust/src",
-                ];
-
-                for &path in &default_paths {
-                    let f = path::Path::new(path);
-                    if f.exists() {
-                        env::set_var("RUST_SRC_PATH", path);
-                        return Ok(())
-                    }
-                }
 
-                Err(RustSrcPathError::Missing)
+    if let Ok(ref srcpaths) = env::var("RUST_SRC_PATH") {
+         if !srcpaths.is_empty() {
+            for path in srcpaths.split(PATH_SEP) {
+                return validate_rust_src_path(path::PathBuf::from(path));
             }
         }
+    };
+
+    if let Some(path) = check_rust_sysroot() {
+        return validate_rust_src_path(path);
+    };
+
+    let default_paths = [
+        "/usr/local/src/rust/src",
+        "/usr/src/rust/src",
+    ];
+
+    for path in default_paths.iter() {
+        if let Ok(path) = validate_rust_src_path(path::PathBuf::from(path)) {
+            return Ok(path);
+        }
+    }
+
+    return Err(RustSrcPathError::Missing)
+}
+
+fn validate_rust_src_path(path: path::PathBuf) -> ::std::result::Result<path::PathBuf, RustSrcPathError> {
+    if !path.exists() {
+        Err(RustSrcPathError::DoesNotExist(path.to_path_buf()))
+    } else if !path.join("libstd").exists() {
+        Err(RustSrcPathError::NotRustSourceTree(path.join("libstd")))
+    } else {
+        Ok(path.to_path_buf())
+    }
+}
+
+#[cfg(test)]
+lazy_static! {
+    static ref TEST_SEMAPHORE: ::std::sync::Mutex<()> = ::std::sync::Mutex::new(());
+}
+
+#[test]
+fn test_get_rust_src_path_env_ok() {
+    use std::env;
+
+    let _guard = TEST_SEMAPHORE.lock().unwrap();
+
+    let original = env::var_os("RUST_SRC_PATH");
+    if env::var_os("RUST_SRC_PATH").is_none() {
+        env::set_var("RUST_SRC_PATH", check_rust_sysroot().unwrap());
+    }
+    assert!(get_rust_src_path().is_ok());
+
+    match original {
```

Comment:
Nitpick: You could create a struct, that stores the original RUST_SRC_PATH and implement Drop. If drop is called, you would reset to the original env var. This could reduce some code duplication in the tests :)